### PR TITLE
Action exception handling on the server

### PR DESCRIPTION
**Pull request by hendrikebbers**
This PR provides Action exception handling on the server. The idea is based on several approaches that JBOSS uses in the JBoss JavaEE APIs (see https://svn.apache.org/repos/infra/websites/production/deltaspike/content/retired/exceptions.html).

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/784)

### CODE REVIEW

**Comment by hendrikebbers** [comment]
Fix for https://github.com/canoo/dolphin-platform/issues/761

**Comment by hendrikebbers** [comment]
This contains only the public API definition at the moment. After a review I will add the implementation / support in DP

**Comment by hendrikebbers** [comment]
@blackdrag ping ping :)

**Comment by blackdrag** [comment]
I see an annotation definition and an exception wrapper named ActionExceptionEvent. There is no usage, no tests, nothing. What exactly am I supposed to review then?

**Comment by hendrikebbers** [comment]
@blackdrag The idea was to define only the public API first (That why this is WIP) and discuss it with you. I tried to describe the usage completely in the JavaDoc.
I think unitTests do not make sense at this state but I can add a (non working) sample that shows how the API should be used. What do you think?

**Comment by blackdrag** [comment]
ok, then let us pass this and go to the implementation ;)

**Comment by blackdrag** [comment]
*[platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java, line 22 at r2](https://reviewable.io:443/reviews/canoo/dolphin-platform/784#-L2UWAnH0L3MFN4_Pb-T:-L2UWAnH0L3MFN4_Pb-U:b-56w93k) ([raw file](https://github.com/canoo/dolphin-platform/blob/8f990898e16e840332df11a54959e2f3dc1ecb14/platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java#L22)):*
> ```Java
>     private final static Logger LOG = LoggerFactory.getLogger(ActionErrorHandler.class);
> 
>     public <T extends Throwable> boolean handle(final T throwable, final Object controller, final String controllerName, final String actionName) {
> ```

what will happen with the original exception if there is an exception in a handle method. In what thread will the exception from the handle method be in?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/784)*

**Comment by blackdrag** [comment]
*[platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java, line 39 at r2](https://reviewable.io:443/reviews/canoo/dolphin-platform/784#-L2UWWc0jMRJxS58TS8D:-L2UWWc0jMRJxS58TS8E:b-cmwl8n) ([raw file](https://github.com/canoo/dolphin-platform/blob/8f990898e16e840332df11a54959e2f3dc1ecb14/platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java#L39)):*
> ```Java
>                 handler.accept(exceptionEvent);
>             } catch (Exception e) {
>                 LOG.error("Error in calling exception handler for error of type '" + throwable.getClass() +
> ```

here we take  care of exceptions of handlers, but the requireNonNull above may cause the same problems

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/784)*

**Comment by blackdrag** [comment]
Reviewed 1 of 2 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/784)*

**Comment by hendrikebbers** [comment]
*[platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java, line 39 at r2](https://reviewable.io:443/reviews/canoo/dolphin-platform/784#-L2UWWc0jMRJxS58TS8D:-L2UcWm1VBNQQpsmLlMn:b-gci5ha) ([raw file](https://github.com/canoo/dolphin-platform/blob/8f990898e16e840332df11a54959e2f3dc1ecb14/platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java#L39)):*
<details><summary><i>Previously, blackdrag (Jochen Theodorou) wrote…</i></summary><blockquote>

here we take  care of exceptions of handlers, but the requireNonNull above may cause the same problems
</blockquote></details>

If the requireNonNull on the top throw an exception we have a general problem in how this method called and error handling is not possible. Anyway such an exception will be logged on the server and the handling call on the client will end as failed. 

The function checks for errors that happen in handlers. But to be true: What would you do if the functionality that calls the handlers is wrong? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/784)*

**Comment by hendrikebbers** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java, line 22 at r2](https://reviewable.io:443/reviews/canoo/dolphin-platform/784#-L2UWAnH0L3MFN4_Pb-T:-L2Ud-XwPuWI1ovO0yhW:bd9ycja) ([raw file](https://github.com/canoo/dolphin-platform/blob/8f990898e16e840332df11a54959e2f3dc1ecb14/platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java#L22)):*
<details><summary><i>Previously, blackdrag (Jochen Theodorou) wrote…</i></summary><blockquote>

what will happen with the original exception if there is an exception in a handle method. In what thread will the exception from the handle method be in?
</blockquote></details>

In what thread will the exception from the handle method be in <- What do you mean? Since this PR is still WIP the handle method is called nowhere. The plan is to call it in the thread that handles the client request that calls the action. Something like:

try {
    callAction();
} catch(Exception e) {
    errorHandler.handle(......);
}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/784)*

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java [hunk 1]
```diff
@@ -80,19 +121,21 @@ public Void run() {
...
-    public static void invokePrivileged(final Method method, final Object instance, final Object... args) {
+    public static <T> T invokePrivileged(final Method method, final Object instance, final Object... args) {
         Assert.requireNonNull(method, "method");
         Assert.requireNonNull(instance, "instance");
-        AccessController.doPrivileged(new PrivilegedAction<Void>() {
+        return AccessController.doPrivileged((PrivilegedAction<T>) new PrivilegedAction<Object>() {
             @Override
-            public Void run() {
+            public Object run() {
                 boolean wasAccessible = method.isAccessible();
```
Comment (giftkugel) [inline platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: Add `final` keyword.

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java [hunk 2]
```diff
@@ -80,19 +121,21 @@ public Void run() {
...
                 try {
                     method.setAccessible(true);
-                    method.invoke(instance, args);
-                    return null; // return nothing...
-                } catch (Exception ex) {
-                    throw new IllegalArgumentException("Cannot invoke method '"
+                    return method.invoke(instance, args);
+                } catch (InvocationTargetException ex) {
```
Comment (giftkugel) [inline platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: Add `final` keyword.

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java [hunk 3]
```diff
@@ -80,19 +121,21 @@ public Void run() {
...
+                    throw new DolphinRuntimeException("Error while calling method '"
+                            + method.getName() + "' on instance of type '" + instance.getClass() + "'. Method details: " + method.toGenericString(), ex);
+                } catch (IllegalAccessException ex) {
```
Comment (giftkugel) [inline platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: Add `final` keyword.

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/ArrayUtils2.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+ * limitations under the License.
+ */
+package com.canoo.dp.impl.platform.core.commons.lang;
+
+import java.lang.reflect.Array;
+import java.util.Arrays;
+
+/**
+ * Some extracted methods from Apache commons lang3
+ */
```
[the remark this answers is not in the document]
Comment (giftkugel) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/ArrayUtils2.java]: I think this class contains some extracted methods from Apache Commons ArrayUtils and should remove the dependency to Apache Commons.

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java [hunk 5]
```diff
@@ -286,7 +296,16 @@ public void invokeAction(final String controllerId, final String actionName, fin
                     index++;
                 }
             }
-            ReflectionHelper.invokePrivileged(actionMethod, controller, args.toArray());
+            try {
+                ReflectionHelper.invokePrivileged(actionMethod, controller, args.toArray());
+            } catch (DolphinRuntimeException e) {
```
Comment (giftkugel) [inline platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java]: Add `final` keyword.

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java [hunk 6]
```diff
@@ -286,7 +296,16 @@ public void invokeAction(final String controllerId, final String actionName, fin
...
+                if(e.getCause() instanceof InvocationTargetException) {
+                    final InvocationTargetException invocationTargetException = (InvocationTargetException) e.getCause();
+                    final Throwable internalException = invocationTargetException.getCause();
+                    actionErrorHandler.handle(internalException, controller, ControllerUtils.getControllerName(controllerClass), actionName);
+                }
+                throw e;
+            }
         } catch (InvokeActionException e) {
```
Comment (giftkugel) [inline platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java]: Add `final` keyword.

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java [hunk 7]
```diff
@@ -286,7 +296,16 @@ public void invokeAction(final String controllerId, final String actionName, fin
...
           throw e;
         } catch (Exception e) {
```
Comment (giftkugel) [inline platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java]: Add `final` keyword.

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        final List<Consumer<ActionExceptionEvent<T>>> consumers = new ArrayList<>();
+        consumers.addAll(getConsumersForTypeInController(throwable.getClass(), controller));
+        LOG.debug("Found {} handlers to handle action exception of type {}", consumers.size(), throwable.getClass());
+
+        final Iterator<Consumer<ActionExceptionEvent<T>>> iterator = consumers.iterator();
+        while (!exceptionEvent.isAborted() && iterator.hasNext()) {
+            final Consumer<ActionExceptionEvent<T>> handler = iterator.next();
+            try {
+                handler.accept(exceptionEvent);
+            } catch (Exception e) {
```
Comment (giftkugel) [inline platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java]: Add `final` keyword.

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                LOG.error("Error in calling exception handler for error of type '" + throwable.getClass() +
+                        "' thrown by action '" + exceptionEvent.getActionName() +
+                        "' of controller '" + exceptionEvent.getControllerName() + "'!", e);
+            }
+        }
+        return exceptionEvent.isAborted();
+    }
+
+    private <T extends Throwable> List<Consumer<ActionExceptionEvent<T>>> getConsumersForTypeInController(Class<? extends Throwable> throwableClass, Object controller) {
```
Comment (giftkugel) [inline platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java]: Add `final` keywords.

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/ArrayUtils2.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/ArrayUtils2.java]: that is right

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/Assert.java [hunk 10]
```diff
@@ -114,6 +115,20 @@ public static boolean isBlank(String str) {
         return collection;
     }
 
+    public static <T> T[] requireNonNullEntries(final T[] array, String argumentName) {
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/Assert.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/Assert.java [hunk 11]
```diff
@@ -114,6 +115,20 @@ public static boolean isBlank(String str) {
...
+        requireNonNull(array, "array");
+        requireNonNullEntries(Arrays.asList(array), argumentName);
+        return array;
+    }
+
+    public static <T> T[] requireNotEmpty(final T[] array, String argumentName) {
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/Assert.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java [hunk 12]
```diff
@@ -36,10 +43,44 @@
...
+        wrapperPrimitiveMap.put(Double.class, Double.TYPE);
+        wrapperPrimitiveMap.put(Float.class, Float.TYPE);
+        wrapperPrimitiveMap.put(Void.class, Void.TYPE);
+    }
+
 
     private ReflectionHelper() {
     }
 
+    public static Optional<Class<?>> getPrimitiveType(Class<?> cls) {
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java [hunk 13]
```diff
@@ -36,10 +43,44 @@
...
+        Assert.requireNonNull(cls, "cls");
+        return Optional.ofNullable(wrapperPrimitiveMap.get(cls));
+    }
+
+    public static Optional<Class<?>> getWrapperClass(Class<?> cls) {
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java [hunk 14]
```diff
@@ -138,6 +181,11 @@ public static Field getInheritedDeclaredField(final Class<?> type, final String
         return result;
     }
 
+    public static Optional<Method> getMethod(final Class<?> type, final String name, Class... paramTypes) {
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java [hunk 15]
```diff
@@ -167,4 +215,75 @@ public static boolean isPrimitiveNumber(final Class<?> cls) {
...
+    public static ParameterizedType toParameterizedType(final Type type) {
+        Assert.requireNonNull(type, "type");
+        if (isParameterizedType(type)) {
+            return (ParameterizedType) type;
+        } else {
+            throw new IllegalArgumentException("The given type is not a ParameterizedType. Type: " + type);
+        }
+    }
+
+    public static boolean isParameterizedType(Type type) {
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/TypeUtils.java [hunk 16]
```diff
@@ -0,0 +1,1418 @@
...
+            // just comparing two classes
+            return toClass.isAssignableFrom((Class<?>) type);
+        }
+
+        if (type instanceof ParameterizedType) {
+            // only have to compare the raw type to the class
+            return isAssignable(getRawType((ParameterizedType) type), toClass);
+        }
+
+        // *
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/TypeUtils.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java (hunk 1 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java (hunk 2 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java (hunk 3 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/ReflectionHelper.java]: done

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java (hunk 5 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java]: done

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java (hunk 6 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java]: done

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java (hunk 7 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/controller/ControllerHandler.java]: done

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java (hunk 8 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java]: done

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java (hunk 9 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/error/ActionErrorHandler.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/TypeUtils.java [hunk 17]
```diff
@@ -0,0 +1,1418 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Some extracted methods from Apache commons lang3
+ */
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/TypeUtils.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/ArrayUtils2.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/commons/lang/ArrayUtils2.java]: done

**Pull request merged**